packages/core/src/render3/util/injector_discovery_utils.ts

```diff
@@ -390,6 +392,43 @@ export function getInjectorProviders(injector: Injector): ProviderRecord[] {
   throwError('getInjectorProviders only supports NodeInjector and EnvironmentInjector');
 }
 
+/**
+ *
+ * Given an injector, this function will return
+ * an object containing the type and name of the injector.
+ *
+ * |              | type        | source                                                      |
+ * |--------------|-------------|-------------------------------------------------------------|
+ * | NodeInjector | Element     | DOM element that created this injector                      |
+ * | R3Injector   | Environment | `injector.source`                                           |
+ * | NullInjector | Null        | "Null Injector"                                             |
+ *
+ * @param injector the Injector to get metadata for
+ * @returns an object containing the type and name of the given injector. If the injector metadata
+ *     cannot be determined, returns null.
+ */
+export function getInjectorMetadata(injector: Injector): {type: string; source: RElement | string}|
+    null {
+  if (injector instanceof NodeInjector) {
+    const lView = getNodeInjectorLView(injector);
+    const tNode = getNodeInjectorTNode(injector)!;
+    assertTNodeForLView(tNode, lView);
+    assertDefined(lView[tNode.index][HOST], 'Could not find node in element view.');
+
+    return {type: 'Element', source: lView[tNode.index][HOST]};
+  }
+
+  if (injector instanceof R3Injector && injector.source !== null) {
+    return {type: 'Environment', source: injector.source};
```

Comment:
I think it's pretty common to not have a source info on R3Injector. Do we still want to return a metadata object in this case to indicate that it's an environment injector? 
```suggestion
  if (injector instanceof R3Injector) {
    return {type: 'Environment', source: injector.source ?? null};
```

Reply:
Good point, added